### PR TITLE
TRAC-1159: Scroll depth refactored

### PR DESCRIPTION
**Pull request by mehakraza**

### CODE REVIEW

File: extensions/doPlugins_global.js [hunk 1]
```diff
@@ -352,7 +352,85 @@ function setPageSourceForCheckout (s) {
...
+    },
+
+    getPagePremiumStatus: function () {
+        const status = window.utag.data.is_status_premium || window.utag.data.page_isPremium
+            || window.utag.data.screen_isPremium;
+        return status ? status + ' : ' : '';
+    },
+
+    isPagenameDefined: function (s) {
+        return !!s.pageName && s.pageName !== '';
```
Comment (HeikoKampe) [inline extensions/doPlugins_global.js:381]: Additional check for empty string is not necessary. !!s.pageName covers empty strings.

File: extensions/doPlugins_global.js [hunk 2]
```diff
@@ -352,7 +352,85 @@ function setPageSourceForCheckout (s) {
...
+                s.prop63 = s._ppvHighestPixelsSeen;
+                s.prop64 = Math.round(s._ppvInitialPercentViewed / 10) * 10;
+                s.prop65 = Math.round(s._ppvHighestPercentViewed / 10) * 10;
+                const event45 = 'event45=' + Math.round(s._ppvInitialPercentViewed / 10) * 10;
+                const event46 = 'event46=' + Math.round(s._ppvHighestPercentViewed / 10) * 10;
+                s.events = s.apl(s.events, event45, ',', 1);
+                s.events = s.apl(s.events, event46, ',', 1);
+            }
+        }
+    },
```
Comment (HeikoKampe) [inline extensions/doPlugins_global.js:430]: The way functions and conditions are nested makes it a bit hard to read and understand. Here is an idea of a simpler structuring:

```
setScrollDepthData: function (s) {
        if (s.pageName) {
            const prevPage = this.setPreviousPage(s);
            if (prevPage) {
                s.getPercentPageViewed(prevPage);
                this.setData(s);
            }
        }
    }
```

File: extensions/doPlugins_global.js [hunk 3]
```diff
@@ -352,7 +352,85 @@ function setPageSourceForCheckout (s) {
...
+                s._prevPage = doc_type + ' : ' + page_is_premium + page_id + ' : ' + page_channel;
+            } else {
+                s._prevPage = s.pageName;
+            }
+        }
+    },
+
+    setScrollDepthData: function (s) {
+        this.setPreviousPage(s);
+        if (this.isPrevPageDefined(s)) {
```
Comment (HeikoKampe) [inline extensions/doPlugins_global.js:414]: I am big fan of wrapping complicated operators of an condition in a utility function. But if it is a simple truthy/falsy check it adds no value.

File: extensions/doPlugins_global.js [hunk 4]
```diff
@@ -404,6 +483,7 @@ s.doPluginsGlobal = function (s) {
     s.eVar184 = new Date().getHours().toString();
     s.eVar181 = new Date().getMinutes().toString();
     s.eVar185 = window.utag.data.myCW || '';
+    s._scrollDepthObj.setScrollDepthData(s);
```
Comment (HeikoKampe) [inline extensions/doPlugins_global.js:486]: The function s.getPercentPageViewed() uses a custom cookies for storing the scroll-depth data. This is only allowed after user has given consent. Therefore a consent check might be needed here. We should discuss this with the team.

**Review by HeikoKampe — CHANGES_REQUESTED** [review]

File: tests/doplugins/doplugins_scroll_depth.test.js [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+
+    describe('isPagenameDefined', () => {
+        
+        it('should return false if pageName is not defined or empty', () => {
+            const sObject = {
+                ...doPlugins_global.s,
+            };
```
Comment (HeikoKampe) [inline tests/doplugins/doplugins_scroll_depth.test.js:21]: Making a copy of the sObject can be moved to the beforeEach hook

File: extensions/doPlugins_global.js (hunk 1 above)
Comment (mehakraza) [reply extensions/doPlugins_global.js:381]: Thanks for pointing out. Updated.

File: extensions/doPlugins_global.js (hunk 3 above)
Comment (mehakraza) [reply extensions/doPlugins_global.js:414]: I agree. Moved the check in if condition.

File: extensions/doPlugins_global.js (hunk 4 above)
Comment (mehakraza) [reply extensions/doPlugins_global.js:486]: The consent check was present inside the scroll depth old implementation in some places and it was decided that it is not necessary there.

File: extensions/doPlugins_global.js (hunk 2 above)
Comment (mehakraza) [reply extensions/doPlugins_global.js:421]: This data is only set once here in the scroll depth context so a separate function would not serve the purpose of reusability but if this implementation decreases readability then there's no harm in moving it to a separate function. Updated in the last commit.

File: extensions/doPlugins_global.js (hunk 4 above)
Comment (HeikoKampe) [reply extensions/doPlugins_global.js]: :+1:

File: tests/doplugins/doplugins_scroll_depth.test.js (hunk 5 above)
Comment (mehakraza) [reply tests/doplugins/doplugins_scroll_depth.test.js]: Updated.

File: tests/doplugins/doplugins_scroll_depth.test.js [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+            jest.spyOn(s._scrollDepthObj, 'getPageId').mockReturnValue('test_pageId');
+            jest.spyOn(s._scrollDepthObj, 'getPageChannel').mockReturnValue('test_pageChannel');
+            jest.spyOn(s._scrollDepthObj, 'getPagePremiumStatus').mockReturnValue('test_is_page_premium_yes : ');
+        
+            s._scrollDepthObj.setPreviousPage(s);
+
+            expect(s._prevPage).toBe('test_docType : test_is_page_premium_yes : test_pageId : test_pageChannel');
+        });
+
+        it('should set the right _prevPage if docType is article or video and page is not premium', () => {
```
Comment (HeikoKampe) [inline tests/doplugins/doplugins_scroll_depth.test.js:212]: As we discussed ... Functionality is implemented and tested in other unit.

File: tests/doplugins/doplugins_scroll_depth.test.js [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        });
+
+    });
+
+    describe('setScrollDepthProperties', () => {
+
+        it('should call setPreviousPage and getPercentPageViewed if pageName is defined', () => {
+            s.pageName = 'test_pageName';
+            const setPrevPage = jest.spyOn(s._scrollDepthObj, 'setPreviousPage');
+            jest.spyOn(s, 'getPercentPageViewed').mockImplementation(jest.fn());
```
Comment (HeikoKampe) [inline tests/doplugins/doplugins_scroll_depth.test.js:231]: MockImplementation creates internally an empty function. Passing in jest.fn() is therefore not needed. 

jest.spyOn(s, 'getPercentPageViewed').mockImplementation(jest.fn());

File: extensions/doPlugins_global.js [hunk 8]
```diff
@@ -357,6 +357,73 @@ s._campaignObj = {
...
+        s.eVar33 = s._ppvPreviousPage;
+        s.prop61 = s._ppvPreviousPage;
+        s.prop62 = s._ppvInitialPercentViewed;
+        s.prop63 = s._ppvHighestPixelsSeen;
+        s.prop64 = Math.round(s._ppvInitialPercentViewed / 10) * 10;
+        s.prop65 = Math.round(s._ppvHighestPercentViewed / 10) * 10;
+        const event45 = 'event45=' + Math.round(s._ppvInitialPercentViewed / 10) * 10;
+        const event46 = 'event46=' + Math.round(s._ppvHighestPercentViewed / 10) * 10;
+        s.events = s.apl(s.events, event45, ',', 1);
+        s.events = s.apl(s.events, event46, ',', 1);
```
Comment (HeikoKampe) [inline extensions/doPlugins_global.js:413]: Here we should use our new way of setting the events property:
s._eventsObj.addEvent(event45);

**Review by HeikoKampe — APPROVED** [review]
Only a view minor issues.

File: extensions/doPlugins_global.js [hunk 9]
```diff
@@ -357,6 +357,73 @@ s._campaignObj = {
...
+
+    getPagePremiumStatus: function () {
+        const status = window.utag.data.is_status_premium || window.utag.data.page_isPremium
+            || window.utag.data.screen_isPremium;
+        return status ? status + ' : ' : '';
+    },
+
+    isDocTypeArticleOrVideo: function () {
+        const doc_type = this.getDocType();
+        return !!doc_type && (doc_type === 'article' || doc_type === 'video');
```
[the remark this answers is not in the document]
Comment (mehakraza) [reply extensions/doPlugins_global.js]: Updated

**Pull request merged**